packages/fleet/lib/utils/locking.ts

```diff
@@ -0,0 +1,45 @@
+import type { Knex } from 'knex';
+import { Ok, Err, stringToHash } from '@nangohq/utils';
+import { setTimeout } from 'node:timers/promises';
+import type { Result } from '@nangohq/utils';
+import { FleetError } from './errors.js';
+
+export async function withPgLock({
+    db,
+    lockKey,
+    fn,
+    onTimeout = async () => {},
+    timeoutMs = 60000
+}: {
+    db: Knex;
+    lockKey: string;
+    fn: () => Promise<Result<void>>;
+    onTimeout?: () => Promise<void>;
+    timeoutMs?: number;
+}): Promise<Result<void>> {
+    let trx: Knex.Transaction | undefined = undefined;
+    try {
+        trx = await db.transaction();
+        const { rows } = await trx.raw<{ rows: { lock: boolean }[] }>(`SELECT pg_try_advisory_xact_lock(?) as lock`, [stringToHash(lockKey)]);
+
+        if (!rows?.[0]?.lock) {
+            await trx.rollback();
+            await setTimeout(1000);
+            return Ok(undefined);
```

Comment:
Why not returning an error if higher up you are waiting 1s on error?

Reply:
you are right. I am doing twice the same thing.